include/puara/utils/blobDetector.h

```diff
@@ -0,0 +1,130 @@
+ * to ensure there is no race conditions.
+ *
+ * @details
+ * The struct maintains internal arrays to store information about detected blobs, including their
+ * start positions, sizes, and centers. It also tracks blob positions from the previous invocation
+ * of the detection function to compute movement values.
+ *
+ * ## Key Functionalities:
+ * - **Blob Detection:** Detects contiguous regions of `1`s in a binary input array.
+ * - **Movement Calculation:** Computes the positional changes (movement) of blobs compared to
+ *   their last positions.
+ * - **Data Management:** Maintains internal state between function calls to facilitate movement
+ *   tracking.
+ *
+ * @note
+ * - The number of blobs processed is limited by `maxNumBlobs` (default is `4`).
+ * - If the input contains more blobs than `maxNumBlobs`, the additional blobs are ignored.
+ */
+struct BlobDetector
+{
+  /** The maximum number of blobs that the algorithm should detect. */
+  static const int maxNumBlobs = 4;
+
+  /** The start index of detected blobs. */
+  int blobStartPos[maxNumBlobs]{};
+
+  /** The cached start index of detected blobs, from the previous time detect1D
+    * was called.
+    */
+  int lastState_blobPos[maxNumBlobs]{};
+
+  /** "size" (amount of stripes) of each blob */
+  int blobSize[maxNumBlobs]{};
+
+  /** shows the "center"(index)of each blob */
+  float blobCenter[maxNumBlobs]{};
+
+  /** amount of detected blobs */
+  int blobAmount{};
+
+  /**
+    * @brief Detects contiguous regions (blobs) of `1`s in a 1D binary array and computes their movement.
+    *
+    * This function identifies blobs in the input binary array `touchArray`, calculates their start
+    * positions, sizes, and centers, and returns the movement of the blobs compared to their positions
+    * from the previous function call.
+    *
+    * @param touchArray Pointer to the 1D binary array representing touch data. Each element is expected to be 0 or 1.
+    * @param size The size of the `touchArray`.
+    * @return A vector of integers representing the movement of each blob's start position since the
+    *         last invocation of `detect1D`. The size of the returned vector is `maxNumBlobs`.
+    *
+    * @note
+    * - The function updates the global variables `blobStartPos`, `blobSize`, `blobCenter`,
+    *   and `lastState_blobPos`.
+    * - The number of blobs detected is limited by `maxNumBlobs`.
+    * - If the number of blobs exceeds `maxNumBlobs`, additional blobs are ignored.
+    *
+    * @warning
+    * - Ensure that `touchArray` has at least `size` elements to avoid out-of-bounds access.
+    * - The function relies on external global variables (`blobStartPos`, `blobSize`, `blobCenter`,
+    *   `lastState_blobPos`, `maxNumBlobs`, `blobAmount`). Ensure they are initialized appropriately
+    *   before calling the function.
+    */
+  std::vector<int> detect1D(const int* const touchArray, const int size)
+  {
+    blobAmount = 0;
```

Comment:
maybe we should protect this function with `if(size >= maxBlobSize) return {};` as right now it would crash due to the various array accesses

Reply:
ah, in fact `size` has nothing to do with `maxBlobSize`, and it wasn't immediately clear to me so I'll try to clarify it in the code. `size` here, which I think I'll rename to `touchArraySize` is the number of touch sensors in the array, which we expect to be more than the number of blobs detected in that array of touch sensors

Reply:
or better yet I'll make it a `small_vector` since we're already doing that as the return type

Reply:
ohh ok

Reply:
Please check my comment on multi versions of the gestures. Since we are separating 1D_blob_detection and the feature extraction, multibrush and multirub should not live in the touch class anymore. They only make sense in the context of touch arrays, and the touch gestures will become 1 DoF generic gestures (taking a float/double as input)